### ISSUE TITLE
integration guide

### ISSUE THREAD

**Issue by yushiangw** [swistakm/pyimgui#125]
Hi,
I am wondering that the requirement of integrate pyimgui with other opengl libs. 
(e.g. [modernGL ](https://github.com/moderngl/moderngl) )

It seems that one only need to reimplment the `imgui/integrations/opengl.py` and forward keyboard/mouse events in user scripts.

**Comment by swistakm** [comment]
Yes. This is how it should be done. I would be happy to review a contribution for such new integration.

**Comment by einarf** [comment]
I might be looking into adding moderngl integration soon:
https://github.com/moderngl/moderngl_window/issues/38

I don't think we should use the `opengl.ProgrammablePipelineRenderer`, but instead create a `ModernglRenderer` what would also contain its own programmable renderer specific for moderngl. (It would be overkill to require PyOpenGL)

**Comment by einarf** [comment]
### Note about pyglet

pyglet 2.x (under development) will move to 3.3 core contexts, so I'm guessing a `Pyglet2Renderer` would be needed in the near future. That means fixed pipeline is out of the question.

**Comment by einarf** [comment]
I just completed a working moderngl5 integration I will PR shortly. I can add a simple integration guide while I have all this info fresh in memory. Should not take too long. I'd at least cover the standard renderer and the mixin for events.

**Comment by swistakm** [comment]
@einarf Great!

**Comment by frosenek** [comment]
In case anyone is looking for a temporary solution for moderngl-window:

    import imgui
    from imgui.integrations import compute_fb_scale
    from imgui.integrations.opengl import ProgrammablePipelineRenderer
    from moderngl_window.context.base.keys import BaseKeys
    from moderngl_window.context.base.window import MouseButtons
    
    
    class MglwRenderer(ProgrammablePipelineRenderer):
    
        def __init__(self, window):
            imgui.create_context()
            super(MglwRenderer, self).__init__()
            window_size = window.window_size
            viewport_size = window.wnd.viewport_width, window.wnd.viewport_height
    
            self.io.display_size = window_size
            self.io.display_fb_scale = compute_fb_scale(window_size, viewport_size)
    
            self._init_key_map(window)
            self._init_event_rerouting(window)
    
        def _init_key_map(self, window):
            keys = window.wnd.keys
    
            self.KEY_MAP = {
                keys.TAB: imgui.KEY_TAB,
                keys.LEFT: imgui.KEY_LEFT_ARROW,
                keys.RIGHT: imgui.KEY_RIGHT_ARROW,
                keys.UP: imgui.KEY_UP_ARROW,
                keys.DOWN: imgui.KEY_DOWN_ARROW,
                keys.PAGE_UP: imgui.KEY_PAGE_UP,
                keys.PAGE_DOWN: imgui.KEY_PAGE_DOWN,
                keys.HOME: imgui.KEY_HOME,
                keys.END: imgui.KEY_END,
                keys.DELETE: imgui.KEY_DELETE,
                keys.SPACE: imgui.KEY_SPACE,
                keys.BACKSPACE: imgui.KEY_BACKSPACE,
                keys.ENTER: imgui.KEY_ENTER,
                keys.ESCAPE: imgui.KEY_ESCAPE,
                keys.A: imgui.KEY_A,
                keys.C: imgui.KEY_C,
                keys.V: imgui.KEY_V,
                keys.X: imgui.KEY_X,
                keys.Y: imgui.KEY_Y,
                keys.Z: imgui.KEY_Z,
            }
    
            for value in self.KEY_MAP.values():
                self.io.key_map[value] = value
    
        def _init_event_rerouting(self, window):
            self.window_key_event = window.key_event
            window.key_event = self.key_event
    
            self.window_mouse_drag_event = window.mouse_drag_event
            window.mouse_drag_event = self.mouse_drag_event
    
            self.window_mouse_position_event = window.mouse_position_event
            window.mouse_position_event = self.mouse_position_event
    
            self.window_mouse_press_event = window.mouse_press_event
            window.mouse_press_event = self.mouse_press_event
    
            self.window_mouse_release_event = window.mouse_release_event
            window.mouse_release_event = self.mouse_release_event
    
            self.window_mouse_scroll_event = window.mouse_scroll_event
            window.mouse_scroll_event = self.mouse_scroll_event
    
            self.window_resize_event = window.resize
            window.resize = self.resize_event
    
            self.window_unicode_char_entered_event = window.unicode_char_entered
            window.unicode_char_entered = self.unicode_char_entered_event
    
        def render(self, data = None):
            data = data if data else imgui.get_draw_data()
            if data:
                super(MglwRenderer, self).render(data)
    
        def update_mouse_position(self, x: int, y: int):
            self.io.mouse_pos = x, y
    
        def key_event(self, key, action, modifiers):
            if key in self.KEY_MAP:
                if action == BaseKeys.ACTION_PRESS:
                    self.io.keys_down[self.KEY_MAP[key]] = True
                else:
                    self.io.keys_down[self.KEY_MAP[key]] = False
            self.window_key_event(key, action, modifiers)
    
        def mouse_drag_event(self, x: int, y: int, dx: int, dy: int):
            self.update_mouse_position(x, y)
            self.window_mouse_drag_event(x, y, dx, dy)
    
        def mouse_position_event(self, x: int, y: int, dx: int, dy: int):
            self.update_mouse_position(x, y)
            self.window_mouse_position_event(x, y, dx, dy)
    
        def mouse_press_event(self, x: int, y: int, button: int):
            self.update_mouse_position(x, y)
    
            if button == MouseButtons.left:
                self.io.mouse_down[0] = 1
    
            if button == MouseButtons.middle:
                self.io.mouse_down[1] = 1
    
            if button == MouseButtons.right:
                self.io.mouse_down[2] = 1
    
            self.window_mouse_press_event(x, y, button)
    
        def mouse_release_event(self, x: int, y: int, button: int):
            self.update_mouse_position(x, y)
    
            if button == MouseButtons.left:
                self.io.mouse_down[0] = 0
    
            if button == MouseButtons.middle:
                self.io.mouse_down[1] = 0
    
            if button == MouseButtons.right:
                self.io.mouse_down[2] = 0
    
            self.window_mouse_release_event(x, y, button)
    
        def mouse_scroll_event(self, x_offset: float, y_offset: float):
            self.io.mouse_wheel = y_offset
            self.window_mouse_scroll_event(x_offset, y_offset)
    
        def resize_event(self, width: int, height: int):
            self.io.display_size = width, height
            self.window_resize_event(width, height)
    
        def unicode_char_entered_event(self, char: str):
            self.io.add_input_character(ord(char))
            self.window_unicode_char_entered_event(char)

**Comment by einarf** [comment]
The integration for moderngl 5.6 (moderngl-window 2.1): https://github.com/moderngl/moderngl-window/blob/2.1/moderngl_window/integrations/imgui.py
I measured it's about twice as fast as the pyopengl one.

Example:
https://github.com/moderngl/moderngl-window/blob/2.1/examples/integration_imgui.py

**Comment by frosenek** [comment]
@einarf Oh wow that's awesome! I totally didn't see that...

**Comment by einarf** [comment]
You can easily hack it to work for moderngl 5.5.x if you remove/replace the scissoring. (`glScissor` support was added in 5.6).

I've been stalling with my PRs here a bit because we did not manage to release 5.6 in December as we hoped. Hopefully soon!

**Comment by frosenek** [comment]
Yeah i noticed the scissoring but it was very easy to remove. Works like a charm now. Thank you so much again for your awesome work!

**Comment by swistakm** [comment]
@einarf: That's sweet! Maybe you could add some note to the ImGui docs referencing that integration so more people could know about it?

**Comment by einarf** [comment]
I will PR it here very soon including a rough writeup on integrations. Just holding it back a bit because moderngl 5.6 is not out yet,
